tensorflow_addons/activations/rrelu_test.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from absl.testing import parameterized
+
+import numpy as np
+import tensorflow as tf
+from tensorflow_addons.activations import rrelu
+from tensorflow_addons.utils import test_utils
+
+
+def _ref_rrelu(x, lower, upper):
+    return tf.where(x >= 0, x, (lower + upper) * x / 2)
+
+
+def _ref_rrelu_grad(x, alpha, dtype):
+    return tf.where(x >= 0, tf.constant(1, dtype=dtype), alpha)
+
+
+@test_utils.run_all_in_graph_and_eager_modes
+class RreluTest(tf.test.TestCase, parameterized.TestCase):
+    @parameterized.named_parameters(("float16", np.float16),
+                                    ("float32", np.float32),
+                                    ("float64", np.float64))
+    @tf.function
+    def test_rrelu(self, dtype):
+        x = tf.constant([-2.0, -1.0, 0.0, 1.0, 2.0], dtype=dtype)
+        lower = 0.1
+        upper = 0.2
+        result = rrelu(x, lower, upper, training=False)
+        expect_result = _ref_rrelu(x, lower, upper)
+        self.assertAllCloseAccordingToType(result, expect_result)
+
+    @parameterized.named_parameters(("float32", np.float32),
+                                    ("float64", np.float64))
+    @tf.function
+    def test_theoretical_gradients(self, dtype):
+        x = tf.constant([-2.0, -1.0, 0.0, 1.0, 2.0], dtype=dtype)
+        lower = 0.1
+        upper = 0.2
+        for training in [True, False]:
+            with self.subTest(training=training):
+                with tf.GradientTape() as t:
+                    t.watch(x)
+                    result, alpha = rrelu(
+                        x, lower, upper, training=training, with_alpha=True)
+                grad = t.gradient(result, x)
+                expect_grad = _ref_rrelu_grad(x, alpha, dtype)
+                self.assertAllCloseAccordingToType(
+                    grad, expect_grad, atol=1e-4)
```

Comment:
Could we use `tf.test.compute_gradient` to check gradients? Just like what you do in `hardtanh`.

Reply:
The test failed when the ```training``` argument is False and test failed when ```training``` argument is True and data is 0.

Reply:
@fsx950223 Sorry for the late reply. Could you push the codes so that I could test it locally? Thanks.

Reply:
[here](https://github.com/tensorflow/addons/pull/573/commits/9aeb856764efb5cd656e4e2f9bd991655e795878)

Reply:
> @fsx950223 Sorry for the late reply. Could you push the codes so that I could test it locally? Thanks.

How do you debug the code. I can't step into source code anymore since tf2 released.

Reply:
Actually, I could clone your forked repo, checkout to this branch, and then test it.

Reply:
> Actually, I could clone your forked repo, checkout to this branch, and then test it.

Could you step into the tf-nightly source code when you are debugging?

Reply:
What's wrong with tf-nightly? Mind providing more information? Thanks

Reply:
I can't step into pip source code anymore since tf2 released but it works well with tf1.
![Screenshot from 2019-10-20 15-38-25](https://user-images.githubusercontent.com/17592563/67156348-e915d980-f34f-11e9-9832-228ec1e519a0.png)

Reply:
What do you mean "cannot step into pip source code anymore since tf2"? Does it mean that you cannot install tf-nightly after tf2 release? Nice vscode config BTW.

Reply:
Edit: I successfully re-configured the environment with `./configure.sh` on the docker early today.

Reply:
It seems something wrong with my vscode config. Thank you for your help.

Reply:
The problem is that I can't go to code definition.
![Screenshot from 2019-10-20 17-49-01](https://user-images.githubusercontent.com/17592563/67157689-1a97a080-f362-11e9-8cd5-4a67fcbb47a4.png)
I tried it in Vscode and Pycharm, but neither of them worked.

Reply:
I believe _compute_numeric_jacobian is not suitable for this test case.Because the activcation has different gradients around 0. Please reference[ Relu test case ](https://github.com/tensorflow/tensorflow/blob/66ea3ed9b8cbbbf01b0eabb14e436883895e4bde/tensorflow/python/kernel_tests/relu_op_test.py#L61)

Reply:
> I believe _compute_numeric_jacobian is not suitable for this test case.Because the activcation has different gradients around 0. Please reference[ Relu test case ](https://github.com/tensorflow/tensorflow/blob/66ea3ed9b8cbbbf01b0eabb14e436883895e4bde/tensorflow/python/kernel_tests/relu_op_test.py#L61)

I think it's ok to modify the input value to avoid non-smooth part. Please see relu's gradient check.
https://github.com/tensorflow/tensorflow/blob/66ea3ed9b8cbbbf01b0eabb14e436883895e4bde/tensorflow/python/kernel_tests/relu_op_test.py#L123

Reply:
> The problem is that I can't go to code definition.
> ![Screenshot from 2019-10-20 17-49-01](https://user-images.githubusercontent.com/17592563/67157689-1a97a080-f362-11e9-8cd5-4a67fcbb47a4.png)
> I tried it in Vscode and Pycharm, but neither of them worked.

Ohoh. Got it. But could the code execute successfully if ide can not find the definition of functions? Sorry that I usually use vim so that I cannot understand what you talk about for a while... Thanks!